lib/fluent/plugin/in_tail.rb

```diff
@@ -669,6 +710,51 @@ def on_timer
       end
     end
 
+    class TailThread
+      class Pool
+        def initialize(max_size, &session)
+          @max_size = max_size
+          @queue = Queue.new
+          @threads = []
+          @running = false
+          session.call(self)
+          @running = true
+        ensure
+          stop
+        end
+
+        def run(&task)
+          @queue.push(task)
+          @threads << create_thread if @threads.size < @max_size
+        end
+
+        def stop
+          return unless @running
+          @running = false
+
+          until @queue.num_waiting == @threads.size
+            sleep 0.01
+          end
+          begin
+            Timeout.timeout(1) do
+              @threads.join
```

Comment:
Do you intent `@threads.each {|th| th.join}`?
`@threads.join` is `Array`'s join so this code doesn't make sense (will return a joined string).

Reply:
Fixed in https://github.com/fluent/fluentd/pull/3185/commits/a03e43cad542af3d0a556e0f0dd0cd5ee2440d59.